tiledb/sm/filesystem/mem_filesystem.cc

```diff
@@ -0,0 +1,357 @@
+
+  std::string get_name() const {
+    return name_;
+  }
+
+  virtual bool is_dir() const = 0;
+  virtual bool is_file() const = 0;
+  virtual std::vector<std::string> ls(const std::string& dir) const = 0;
+  virtual bool has_child(const std::string& child) const = 0;
+
+  // check if map instead
+  std::unordered_map<std::string, FSNode*> children_;
+
+ protected:
+  std::string name_;
+};
+
+class MemFilesystem::File : public MemFilesystem::FSNode {
+ public:
+  File(const std::string& name)
+      : FSNode(name){};
+
+  ~File() {
+    free(data_);
+  };
+
+  bool is_dir() const override {
+    return false;
+  }
+
+  bool is_file() const override {
+    return true;
+  }
+
+  std::vector<std::string> ls(const std::string& dir) const override {
+    return {dir + name_};
+  }
+
+  bool has_child(const std::string&) const override {
+    return false;
+  }
+
+  uint64_t get_size() const {
+    return size_;
+  }
+
+  // TODO: size is in number of bytes
+  Status append(const void* data, const uint64_t nbytes) {
+    if ((data == nullptr) || nbytes == 0) {
+      return LOG_STATUS(Status::MemFSError(std::string(
+          "Wrong input buffer or size when writing to file: " + name_)));
+    }
+
+    /* if the file was empty */
+    if (data_ == nullptr) {
+      data_ = malloc(nbytes);
+      if (data_ == nullptr) {
+        return LOG_STATUS(Status::MemFSError(
+            std::string("Out of memory, cannot write to file: " + name_)));
+      }
+
+      size_ = nbytes;
+    } else {
+      /* if the file already had some data */
+      void* new_data;
+      new_data = realloc(data_, size_ + nbytes);
+
+      if (new_data == nullptr) {
+        return LOG_STATUS(Status::MemFSError(std::string(
+            "Out of memory, cannot append new data to file: " + name_ +
+            "with size " + "size_")));
+      }
+
+      std::memcpy((char*)new_data + size_, data, nbytes);
+      data_ = new_data;
+      size_ += nbytes;
+    }
+
+    return Status::Ok();
+  }
+
+  Status read(
+      const uint64_t offset, void* buffer, const uint64_t nbytes) const {
+    assert(buffer);
+    if (offset + nbytes > size_)
+      return LOG_STATUS(
+          Status::MemFSError("Cannot read from file; Read exceeds file size"));
+
+    std::memcpy(buffer, (char*)data_ + offset, nbytes);
+    return Status::Ok();
+  }
+
+ private:
+  void* data_;
+  uint64_t size_;
+};
+
+class MemFilesystem::Directory : public MemFilesystem::FSNode {
+ public:
+  Directory(const std::string& name)
+      : FSNode(name){};
+
+  // TODO: add body
+  ~Directory(){};
+
+  bool is_dir() const override {
+    return true;
+  }
+
+  bool is_file() const override {
+    return false;
+  }
+
+  std::vector<std::string> ls(const std::string& dir) const override {
+    std::vector<std::string> names;
+    for (const auto& child : children_) {
+      names.emplace_back(dir + child.first);
+    }
+    /* TODO: check if needed
+    sort(result.begin(), result.end());
```

Comment:
Instead, we could sort the children when inserted into the directory.

Reply:
What would be the benefit of that if we only need them sorted in list operation?

Reply:
I was thinking that the benefit would be that we sort once on insert rather than on every ls, but I guess the better performance would depend entirely on the workload. _However_, now that we've decided to map the children with `unordered_map<string, FSNode*>`, let's leave this here because we would lose the O(1) access in the lookups by moving to a sorted `map<string, FSNode*>`.